language/move-prover/bytecode-to-boogie/src/driver.rs

```diff
@@ -0,0 +1,188 @@
+    pub fn run(&mut self) {
+        self.load_modules(&self.options.mvir_sources);
+        self.add_prelude();
+        self.translate_modules();
+        // write resulting code
+        info!("writing boogie to {}", self.options.output_path);
+        abort_on_error(
+            fs::write(&self.options.output_path, &self.output),
+            "cannot write boogie file",
+        );
+        if !self.options.generate_only {
+            self.call_boogie(&self.options.output_path);
+        }
+    }
+
+    /// Runs translation in a test context. Instead of writing output to a file, returns
+    /// a pair of the prelude and the actual translated code.
+    pub fn run_for_test(&mut self) -> (String, String) {
+        self.load_modules(&self.options.mvir_sources);
+        self.add_prelude();
+        // Extract the prelude from the generated output, as we want to return it in a separate
+        // string.
+        let prelude = std::mem::replace(&mut self.output, String::new());
+        self.translate_modules();
+        (prelude, std::mem::replace(&mut self.output, String::new()))
+    }
+
+    /// Compiles the list of mvir modules and adds them to the state. On errors, this
+    /// reports and aborts. Nothing will be added to the output yet.
+    pub fn load_modules(&mut self, file_names: &[String]) {
+        let address = AccountAddress::default();
+        for (index, file_name) in file_names.iter().enumerate() {
+            info!("analyzing {}", file_name);
+            // Parse module.
+            let code = abort_on_error(fs::read_to_string(file_name), "cannot read mvir file");
+            let parsed_module = abort_on_error(parse_module(&code), "mvir parsing errors");
+            let name = parsed_module.name.as_inner().to_string();
+
+            // Extract specifications from module.
+            let specs = self.extract_function_infos(&parsed_module);
+
+            // Compile module.
+            let (compiled_module, source_map) = abort_on_error(
+                compile_module(address, parsed_module, &self.verified_modules),
+                "mvir compilation errors",
+            );
+
+            // Verify module byte code.
+            let verified_module = abort_on_error(
+                VerifiedModule::new(compiled_module)
+                    // As an error message, produce a line separated list of VMStatus.
+                    .map_err(|(_, sv)| sv.iter().map(|s| format!("{:?}", s)).join("\n")),
+                "mvir verification errors",
+            );
+
+            // Store result.
+            self.verified_modules.push(verified_module);
+            self.module_infos.push(ModuleInfo {
+                index,
+                name,
+                function_infos: specs,
+            });
+            self.source_map.push(source_map);
+        }
+    }
+
+    /// Extract function infos from the given parsed module.
+    fn extract_function_infos(&self, parsed_module: &ModuleDefinition) -> Vec<FunctionInfo> {
```

Comment:
Nit: could use `ModuleDefinition` here and save some clones given that the caller always has an owned `ModuleDefinition`.